### ISSUE TITLE
Where is linear fuction?

### ISSUE THREAD

**Issue by jo-kyeongbin** [porcelainruler/Gated-Linear-Activation-Implementation-TF#1]
Hi!
I have tensorflow-gpu==1.14. when I run this code included example code, model summary of results are none.
and, GLU need to linear fuction. but, this code don't have linear fuction.
I wonder I understand this code well. help!

**Comment by porcelainruler** [comment]
Hi,
In Gated Linear Activation, the input is split in two half along the channel, for e.g. in case of an input of dimension (a, b, c), it is sliced into two equal slices of (a, b, c/2) each, let them be s1 and s2. And the output should be s1 * sigmoid(s2) (element-wise multiplication) s1 is the linear part and sigmoid acts as a gate. Since the splitting is around channel dimension the output will have half the dimension along the channels. This leads to faster convergence in training conv-nets and doesn't suffer from vanishing gradient problem like in case of tanH and GTU activation.

And the model summary is none because it is only an activation function like sigmoid, tanh, etc, and doesn't have any trainable parameter. The output of a random tensor (like a 1-dim Conv layer) is shown in the below colab notebook.
Link - https://colab.research.google.com/drive/1m-Kp9AU9c-WmkF4UQNXsMdd-kI3Ap6Ar#scrollTo=gTPmqxzk__CQ

Also, for the detailed study, you can refer to this paper: https://arxiv.org/abs/1612.08083v3
And PyTorch have implementation of glu layer: https://pytorch.org/docs/1.9.0/generated/torch.nn.functional.glu.html

Discussion - https://github.com/pytorch/pytorch/issues/16029

**Comment by jo-kyeongbin** [comment]
Thank you for quick response.

Lastly,

1.
I found input parameter "(self, filters, kernel_size=3, dim=-1, inp_shape=(128, 100, 64))",
but, filters, kernel_size, inp_shape are **not used**, What do these parameters do?
  Is the kernel the weight of a linear function?
  It works fine without inp_shape. What is inp_shape?

2.
I read paper that you recommend, but The link below seems a bit different.
https://medium.com/deeplearningmadeeasy/glu-gated-linear-unit-21e71cd52081

  1) output = linear(input) * linear(sigmoid(input))
  
  2)
      input / 2 --> a , b
      output = linear(input(a)) * linear(input(sigmoid(b)))

Is the second expression correct?

**Comment by porcelainruler** [comment]
Hi,
1. I have written it around 1 year back, back then I was thinking of extending it a bit since TensorFlow doesn't have the glu activation at that time and I want to make it similar to the PyTorch one but doesn't get time. And PyTorch has a similar structure, but they use passed parameters values instead of the input_shape that I have extracted from the passed tensor. Both give the same output.
2. the second expression is correct gated linear unit activation work as putting a gate or a tap in linear input but also half the output channel/features of the tensor. Also, you can find the exact implementation in PyTorch nn.functional.glu. Do read the code either in their Github repo or directly in the python package it will become a lot more clear.

https://leimao.github.io/blog/Gated-Linear-Units/  -> This is a great article. It has cleared a lot of my doubts regarding glu activation.
